### DISCUSSION TITLE
Forward authorized request inside controller/service

### DISCUSSION THREAD

**Discussion by isBrainDed** [dotnet/yarp#2045] [Q&A] [answered]
Hello,

I am developing a .Net Core 7 API which allows users to connect to WS Streams for specific cameras using YARP as a reverse proxy.

I have implemented a proof of concept with Direct forwarding, but I am not impressed with the solution.
What is basically does. Upon request "wss://localhost:7071/camera/livestream/1/rtsp-over-websocket", it finds the Camera with ID 1 and then redirects the request to the Camera's IP, with YARP automatically handling upgrade to the WS, because Cameras also require authentication with the user not knowing the credentials, API also authenticates with the camera and injects the header
```cs
var transformer = new CustomTransformer(); // or HttpTransformer.Default;
var requestOptions = new ForwarderRequestConfig { ActivityTimeout = TimeSpan.FromSeconds(100) };
app.MapForwarder("/camera/livestream/{**catch-all}", "127.0.0.1", requestOptions, transformer);

//// .... ////

private class CustomTransformer : HttpTransformer
{
	public override async ValueTask TransformRequestAsync(HttpContext httpContext, HttpRequestMessage proxyRequest, string destinationPrefix)
	{
		// Copy all request headers
		await base.TransformRequestAsync(httpContext, proxyRequest, destinationPrefix);

		// Customize the query string:
		var queryContext = new QueryTransformContext(httpContext.Request);
		string sanitizedPath = httpContext.Request.Path.Value.Replace("/camera/livestream/", "/");
		var stringSplit = sanitizedPath.Split('/', StringSplitOptions.RemoveEmptyEntries);
		string finalPath = "";

		for (int i = 0; i < stringSplit.Length; i++)
		{
			if (i > 0)
			{
				finalPath += "/" + stringSplit[i];
			}
		}

		foreach (CameraService cameraService in CameraServiceCollection.Instance)
		{
			if (cameraService.Id.ToString() == stringSplit[0])
			{
				proxyRequest.RequestUri = RequestUtilities.MakeDestinationAddress($"http://{cameraService.URL}", finalPath, queryContext.QueryString);
				using (var client = new HttpClient(new HttpClientHandler { Credentials = new NetworkCredential(cameraService.Username, cameraService.Password) }))
				{
					try
					{

						string url = $"http://{cameraService.URL}{finalPath}";
						using (HttpResponseMessage response = await client.GetAsync(url, HttpCompletionOption.ResponseHeadersRead))
						{
							proxyRequest.Headers.Authorization = response.RequestMessage.Headers.Authorization;
						}

					}
					catch (Exception ex) { }
				}
			}

		}
		proxyRequest.Headers.Host = null;
	}
}
```
This is the request that is caught by the MapForwarder and properly redirected
![image](https://user-images.githubusercontent.com/41236638/220616552-80bc6c5c-3fed-4f79-8efd-0ed3a910218f.png)

The issues that I have with the above implementation is that it doesn't "fit" in the Controller/Service model. Because I am not registering it in the controller, I can't use the .Net Identity authentication to see if the request came from an User that has permission to receive the Camera stream, or should we deny the request ( That I could find )

I also have to register the destination IP to 127.0.0.1 because I do not know the amount of cameras that I would have to proxy too and they could change at runtime. ( I noticed it is possible to refresh the configuration at runtime )

Would it be possible to inject the YARP modules in API Controllers and upgrade to WS and proxy in them? That would be the cleanest solution in my opinion

Very rough diagram of design
![image](https://user-images.githubusercontent.com/41236638/220619314-cf593dcc-827e-4e25-86cd-ad5be082dc9d.png)

**Comment by Tratcher** [comment, marked as answer]
You don't need MapForwarder, that's only for the simplest scenarios, you can call the IHttpForwarder directly from a normal controller. Most of that transform logic would move into the controller before invoking the forwarder, which addresses your question about 127.0.0.1.

> I noticed it is possible to refresh the configuration at runtime

This isn't relevant to direct forwarding, you're not reading anything here from config.

**Reply by isBrainDed** [reply]
Hey, 

thank you for the swift reply. I actually got it working with the little nudge that you gave me.

I used the same CustomTransformer as in the original post, but here is the API controller now. 

```cs
[Route("stream/{*path}")]
[HttpGet]
public async Task ForwarderHandler()
{
	var httpClient = new HttpMessageInvoker(new HttpClientHandler()
	{
		UseProxy = false,
		AllowAutoRedirect = false,
		AutomaticDecompression = DecompressionMethods.All,
		UseCookies = false,
	});
	var transformer = new CustomTransformer();
	var requestConfig = new ForwarderRequestConfig { ActivityTimeout = TimeSpan.FromSeconds(100) };

	// Destination prefix seems to be able to be anything but HTTPS?
	var error = await _httpForwarder.SendAsync(HttpContext, "127.0.0.1",
		httpClient, requestConfig, transformer);
	// Check if the operation was successful
	if (error != ForwarderError.None)
	{
		var errorFeature = HttpContext.GetForwarderErrorFeature();
		var exception = errorFeature?.Exception;
	}
}
```

It upgrades to the websocket correctly.

I did note that if you use "https://localhost:10001/Path/Base" or any HTTPS as shown in the Transforms example https://microsoft.github.io/reverse-proxy/articles/transforms.html , it first tries to do CONNECT instead of GET, so the WebSockets never connect. I am guessing because the initial request has to first establish HTTPS connection with the destination

HTTPS destination prefix: 
![image](https://user-images.githubusercontent.com/41236638/220857666-364cb055-191d-4f00-b6fd-24b1666677b0.png)

HTTP destination prefix:
![image](https://user-images.githubusercontent.com/41236638/220857790-9c8205f1-9938-47c7-8325-e70362a93382.png)


One more question, but it is not that important or deal breaking. 
When a WebSocket connection is established and consumed it works fine, but when the WebSocket client suddenly ends  ( page refresh, browser closing ), an exception is sometimes logged in the output but not caught anywhere. Is that just debug information or should we be cleaning that up better somehow?

```
System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..
 ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Net.Http.HttpConnection.ReadAsync(Memory`1 destination)
   at System.Net.Http.HttpConnection.RawConnectionStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.RawConnectionStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at Yarp.ReverseProxy.Forwarder.StreamCopier.CopyAsync(Stream input, Stream output, Int64 promisedContentLength, StreamCopierTelemetry telemetry, ActivityCancellationTokenSource activityToken, Boolean autoFlush, CancellationToken cancellation)
```

**Reply by Tratcher** [reply]
Ah, when you use HTTPS it negotiates HTTP/2, and then attempts WebSockets over HTTP/2 (CONNECT). That's a new feature supported by YARP 2.0 and .NET 7. If that doesn't work it should retry and fall back to HTTP/1.1 (GET). Can you share the debug logs? I'm curious why that retry isn't working.

You can also work around that by specifying version 1.1 in ForwarderRequestConfig (2.0 is the default).

**Reply by Tratcher** [reply]
> One more question, but it is not that important or deal breaking.
When a WebSocket connection is established and consumed it works fine, but when the WebSocket client suddenly ends ( page refresh, browser closing ), an exception is sometimes logged in the output but not caught anywhere. Is that just debug information or should we be cleaning that up better somehow?

The forwarder catches those exceptions and handles them as part of normal operations, you shouldn't need to do anything.

**Reply by isBrainDed** [reply]
> Ah, when you use HTTPS it negotiates HTTP/2, and then attempts WebSockets over HTTP/2 (CONNECT). That's a new feature supported by YARP 2.0 and .NET 7. If that doesn't work it should retry and fall back to HTTP/1.1 (GET). Can you share the debug logs? I'm curious why that retry isn't working.
> 
> You can also work around that by specifying version 1.1 in ForwarderRequestConfig (2.0 is the default).

Yeah, that is my bad. It works correctly, I thought that the cameras had valid HTTPS certificates but they don't. So it did fallback to HTTP/1.1 but the certificate was invalid. Sorry for wasting your time on that.
```
      Unable to proxy the WebSocket using HTTP/2, server does not support HTTP/2. Retrying with HTTP/1.1. Disable HTTP/2 negotiation for improved performance.
info: Yarp.ReverseProxy.Forwarder.HttpForwarder[48]
      Request: An error was encountered before receiving a response.
      System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
       ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: NotSignatureValid, UntrustedRoot
```

Thank you so much for the assistance and the guidance again.
